### PR TITLE
Convert System.Globalization unix calls to QCalls into coreclr

### PR DESCRIPTION
**Pull request by safern**
Relates to: https://github.com/dotnet/runtime/issues/826
Since we now need to include the headers as part of mscorlib in order to define the QCall entry points, I had to use `uint16_t` instead of `UChar`. Per ICU's documentation `UChar` is an unsigned int of 16bits -- since we don't want to include headers from external libraries into coreclr build. 

**Note: marked as no-merge as since this is touching perf sensitive code paths, to measure perf.**

cc: @ericstj

### CODE REVIEW

**Comment by jkotas** [comment]
I think it would be better to compile this separately from vm, without including PAL headers, to avoid collisions with PAL headers.

**Comment by safern** [comment]
> I think it would be better to compile this separately from vm, without including PAL headers, to avoid collisions with PAL headers. Dealing with the PAL and system headers collisions is a losing battle.

Do you mean defining new headers for the functions that are going to be qcalls directly in the vm? Or could you provide an example?

**Comment by safern** [comment]
> Do you mean defining new headers for the functions that are going to be qcalls directly in the vm? Or could you provide an example?

I see, so you actually mean compiling in another partition of the runtime, which one would you suggest?

File: src/coreclr/src/vm/ecalllist.h
Comment (jkotas) [inline src/coreclr/src/vm/ecalllist.h]:
```suggestion
#ifdef TARGET_UNIX
```

**Comment by jkotas** [comment]
>  you actually mean compiling in another partition of the runtime, which one would you suggest?

Yep. It can be the `src\coreclr\src\libraries-native` that you have added in last PR. 

I think you should be able to keep the shim implementation intact, and just add `src\coreclr\src\libraries-native\entrypoints.c` file that looks like this:

```
#include "pal_calendarData.h"
#include "pal_casing.h"
...

#define FCFuncStart(name) extern const void* name[]; const void* name[] = {
#define FCFuncEnd() 0x01 /* FCFuncFlag_EndOfArray */ };

#define QCFuncElement(name,impl) \
    (BYTE*)0x8 /* FCFuncFlag_QCall */, (LPVOID)(impl), (LPVOID)name,

FCFuncStart(gPalGlobalizationNative)
    QCFuncElement("ChangeCase", GlobalizationNative_ChangeCase)
    QCFuncElement("ChangeCaseInvariant", GlobalizationNative_ChangeCaseInvariant)
...
FCFuncEnd()
```

And then have a registration in `ecalllist.h` like:

```
extern "C" const void* gPalGlobalizationNative[];
...
FCClassElement("Globalization", "", gPalGlobalizationNative)
...
```

**Comment by safern** [comment]
Thanks, I will try that!

File: src/libraries/Native/Unix/Common/pal_compiler.h
Comment (jkotas) [inline src/libraries/Native/Unix/Common/pal_compiler.h]: It may be a good idea to explicitly strip exporting of the shim entrypoints when compiling them as part of CoreCLR. They do not need to be exported.

Maybe define a new macro `PALEXPORT` and change all shims to use it. The macro can then be defined as needed depending on the context.

Comment (safern) [reply src/libraries/Native/Unix/Common/pal_compiler.h]: Sounds good.

File: src/libraries/Native/Unix/Common/pal_compiler.h
Comment (safern) [inline src/libraries/Native/Unix/Common/pal_compiler.h]: @jkotas did I understand correctly what you meant here? Or do this needs to always be "hidden" for System.Globalization.Native now that it is linked into coreclr, even when building the implementation?

Comment (safern) [reply src/libraries/Native/Unix/Common/pal_compiler.h]: I think I got what you said, I just changed it to define it as "default" here if not defined and then in GlobalizationNative defined it to "hidden" before importing `pal_compiler.h`

File: src/libraries/Native/Unix/Common/pal_compiler.h
Comment (safern) [reply src/libraries/Native/Unix/Common/pal_compiler.h]: It seems like I can't do this yet for S.G.Native because if so then mono will not find the entrypoints as they're still PInvoking into S.G.Native. We will be able to do this until we link mono into it.

File: src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs
Comment (jkotas) [inline src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs]: It would be best to get rid of the use of ICU version in the test. The tests can simply allow both results as valid, and the ICU versions that the different results apply to can be captured in comment.

Comment (jkotas) [reply src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs]: If you really want to keep testing the ICU version (I do not think it is useful...), I would call he internal CoreLib method that returns the ICU version via private reflection. It will work same way on both CoreCLR and Mono, and be simpler overall.

Comment (safern) [reply src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs]: > It would be best to get rid of the use of ICU version in the test. The tests can simply allow both results as valid, and the ICU versions that the different results apply to can be captured in comment.

Some of the tests skip test data for some versions of ICU because there is ICU fixes for overflows and stuff like that. https://github.com/dotnet/runtime/blob/master/src/libraries/System.Globalization/tests/CompareInfo/CompareInfoTests.IsPrefix.cs#L81

You don't think is worth doing that?

> ICU version via private reflection.

That's a good idea.

File: src/coreclr/src/libraries-native/CMakeLists.txt
Comment (jkotas) [inline src/coreclr/src/libraries-native/CMakeLists.txt]: `FEATURE_GLOBALIZATION` makes it sound like that this is used to enable/disable globalization support in the runtime. It is not the case.

I guess you have added this ifdef to turn the ICU globalization shim on/off. I do not think you need this ifdef. We are working towards having ICU everywhere, and so this ifdef would be eventually enabled everywhere.

For now, I think you can just use `TARGET_UNIX` for the few places - until you do the work to get ICU compiled on Windows too.

File: src/coreclr/src/libraries-native/entrypoints.c
Comment (jkotas) [inline src/coreclr/src/libraries-native/entrypoints.c]: This should be always compiled as C. Do you really need this?

File: src/coreclr/src/libraries-native/CMakeLists.txt
Comment (safern) [reply src/coreclr/src/libraries-native/CMakeLists.txt]: Yeah. This was a leftover of the initial design I had for the entrypoints.c where I was compiling it as part of System.Globalization.Native instead of libraries-native. Since entrypoints.c is potentially going to be shared across different shims, I added this for that, but it is not needed anymore. Thanks.

File: src/coreclr/src/libraries-native/entrypoints.c
Comment (jkotas) [inline src/coreclr/src/libraries-native/entrypoints.c]:
```suggestion
#define FCFuncStart(name) extern const void* name[]; const void* name[] = {
```

And delete the `#ifndef`s above

File: src/coreclr/src/libraries-native/entrypoints.c
Comment (jkotas) [inline src/coreclr/src/libraries-native/entrypoints.c]:
```suggestion
#define FCFuncEnd() (void*)0x01 /* FCFuncFlag_EndOfArray */ };
```

File: src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs
Comment (jkotas) [inline src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs]: There is no single `#if CORECLR` under libraries/common. I think `#if CORECLR` under libraries/common may get confusing very fast. I am wondering whether we should rather recognize `System.Globalization.Native` as a special library in the runtime, and leave this one intact.

It is a single place in dllimport.cpp:

```
    if (pNMD->GetLoaderModule()->IsSystem() && (strcmp(szLibName, "QCall") == 0 || strcmp(szLibName, "System.Globalization.Native"))
    {
        ndirectflags |= NDirectMethodDesc::kIsQCall;
    }
```

File: src/libraries/Native/Unix/System.Globalization.Native/pal_locale.h
Comment (jkotas) [inline src/libraries/Native/Unix/System.Globalization.Native/pal_locale.h]: Would it look better to spit this file into two?

File: src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs
Comment (jkotas) [reply src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs]: > You don't think is worth doing that?

What does that test do on old ICU versions? Crash or just return wrong result?

File: src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs
Comment (safern) [reply src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs]: Sounds reasonable, but wouldn't that make it harder for people when reading the Interop code to realize it is a QCall? Or it doesn't really matter?

File: src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs
Comment (safern) [reply src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs]: Hmm there is only two usages of ICUVersion in the tests, one the one that I pointed above and this one: https://github.com/dotnet/runtime/blob/9999d4b86e0c7b0e7fbc04c24bad346397ef17ff/src/libraries/System.Globalization/tests/System/Globalization/RegionInfoTests.cs#L145 -- which we could make both expected values valid in the tests.

> What does that test do on old ICU versions? Crash or just return wrong result?

I don't know, I'll let @tarekgh answer that, he should know more.

File: src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs
Comment (safern) [reply src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs]: Also, does that mean that when we link this into mono, we would have to special case it in mono runtime as well?

Comment (jkotas) [reply src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs]: > Or it doesn't really matter?

I do not think it matters. When you are reading the code, you typically do not care how it is stitched together - you only care about that when reading, e.g. cmake file that links everything together.

> Also, does that mean that when we link this into mono, we would have to special case it in mono runtime as well?

Yes, I expect that it would be special-cased somehow. I do not know what the most natural solution for Mono would be.

File: src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs
Comment (tarekgh) [reply src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs]: We don't crash but ICU return wrong results. I am still seeing have ICUVersion is very useful for us. Sometimes it helps in diagnosing some raised issues without the need for deep debugging. Another minor reason is when writing the behavior of different ICU version in the test helps us to know the history of the change. This is really useful in some cases when it is needed instead of going back and try to figure out what has changed. 

In short, I don't mind changing the tests but I am still seeing a value keeping the ICU version property in our tests.

File: src/libraries/Native/Unix/Common/pal_compiler.h
Comment (jkotas) [reply src/libraries/Native/Unix/Common/pal_compiler.h]: This looks fine. You may be able to do it by adding:
 ```
// Suppress exporting of the PAL APIs
add_definitions(-DPALEXPORT)
```
to `libraries-native/CMakeLists.txt`.

Comment (safern) [reply src/libraries/Native/Unix/Common/pal_compiler.h]: I see. Hmm, that's a good idea. Let me try that.

File: src/coreclr/src/libraries-native/entrypoints.c
Comment (safern) [reply src/coreclr/src/libraries-native/entrypoints.c]: I don't think so. Will remove.

File: src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs
Comment (safern) [reply src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs]: Ok, sounds good. Then I'll special case this in the runtime directly.

File: src/libraries/Native/Unix/System.Globalization.Native/pal_locale.h
Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_locale.h]: I wanted to avoid touching all the files that imported this header to import the new header. 

Or I can split this file in two and just import the new header in this header ifndef __LIB_NATIVE_ENTRYPOINTS.

Comment (jkotas) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_locale.h]: > I wanted to avoid touching all the files that imported this header 

It is trivial find&replace. I do not see that as a problem.

This header diverges from the style used by all the other headers (it is mixing the entrypoints and internal helpers). I think there is a value in uniformity.

Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_locale.h]: Ok, I'll do that. there is another one which does that as well `pal_errors.h` so I'll do it for that one also.

**Comment by safern** [comment]
Thanks for the reviews 😄 -- I've addressed all feedback.

File: src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs
Comment (jkotas) [inline src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs]: All changes to this file can reverted

Comment (safern) [reply src/libraries/Common/src/Interop/Unix/Interop.Libraries.cs]: 🤦‍♂

File: src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs
Comment (jkotas) [inline src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs]: This should not be inside `class libc`. It has nothing to do with libc. It can be in the place that calls `libc.GetICUVersion` above.

File: src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.h
Comment (jkotas) [inline src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.h]: I thought that you have done the renaming to `pal_locale_internal` so that this is no longer necessary?

File: src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs
Comment (jkotas) [inline src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs]: Instead of IsWindows check, we can just use presence/absence of the GetICUVersion internal method to tell whether the ICU is present (future proofing for when ICU is enabled on Windows); and return null 0.0.0.0 when it is absent.

File: src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.h
Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.h]: `pal_icushim.h` is still the one that brings the `ICUHeaders`. Which we use in all our headers and implementations. However when doing the entrypoints build I define UChar at the bottom of this file. If you want me to that, I could move the ICU header inclusions file to a different header file and do the right things in order to not include config.h nor these headers when doing the entrypoints build.

File: src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs
Comment (safern) [reply src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs]: 👍

File: src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs
Comment (tarekgh) [reply src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs]: Interesting we have this check in the first place in .Unix file. I agree we should have a method telling if ICU is used or not. That what we are going to have anyway when we start enabling ICU for Windows.

---
In reply to: [379152410](https://github.com/dotnet/runtime/pull/32132#discussion_r379152410) [](ancestors = 379152410)

Comment (safern) [reply src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs]: We have the Windows Detection on a .Unix file because TestUtilities is not cross compiled by OS, so the convention is to use runtime checks instead.

File: src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.h
Comment (jkotas) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.h]: I will leave it up to you.

Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.h]: Thanks. I will see how it looks with moving the icu headers imports to a new file. Then I'll decide, anyway, I would still like to use `#ifdef __LIB_NATIVE_ENTRYPOINTS` to define `UChar` to make sure that that type is only defined in our side when building the entrypoints.

Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.h]: At the end I decided to move it to `pal_icushim_internal.h` and get rid of `__LIB_NATIVE_ENTRYPOINTS` so that entrypoints.c and globalization are de-coupled. Thanks

**Review by jkotas — APPROVED** [review]

**Comment by jkotas** [comment]
The test failures seems to be related to this change. Do not merge before fixing them.

**Comment by safern** [comment]
Yes I saw that, will investigate later. It seems to be related to getting the version. Thanks for all your reviews @jkotas!

File: src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs
Comment (jkotas) [inline src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.Unix.cs]:
```suggestion
                    version = (int)methodInfo.Invoke(null, null);
```

File: src/mono/netcore/System.Private.CoreLib/src/LinkerDescriptor/System.Private.CoreLib.xml
Comment (akoeplinger) [inline src/mono/netcore/System.Private.CoreLib/src/LinkerDescriptor/System.Private.CoreLib.xml]: would be nice to move this to the bottom of the file, there is already another test-only dependency there.

Comment (safern) [reply src/mono/netcore/System.Private.CoreLib/src/LinkerDescriptor/System.Private.CoreLib.xml]: I thought it was sorted alphabetically by class. Will update and move it down.

Comment (safern) [reply src/mono/netcore/System.Private.CoreLib/src/LinkerDescriptor/System.Private.CoreLib.xml]: Fixed.

**Pull request merged**
